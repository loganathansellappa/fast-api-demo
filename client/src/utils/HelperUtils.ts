import { toast } from 'react-toastify';

export type ServerConfig = {
    baseUrl: string;
};
export const serverData = () => {
    return {
        baseUrl: process.env.DATA_URL,
    } as ServerConfig;
};

export enum MissionState {
    PreFlight = 'Pre-flight',
    Flight = 'In-Flight',
    PostFlight = 'Post-Flight'
}


export enum MutationType {
    PUT = 'PUT',
    POST = 'POST',
    DELETE = 'DELETE'
}

export const showSuccessToast = (message: string, id: string | number = 1) => {
    toast.dismiss();
    toast.success(message, { toastId: id, autoClose: 500, style: { background: '#5CDD00', color: "white" } });
};

export const showErrorToast = (message: string, id: string | number) => {
    toast.dismiss();
    toast.error(message, { toastId: id, autoClose: 500 });
};
