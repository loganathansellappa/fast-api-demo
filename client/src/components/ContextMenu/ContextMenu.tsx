import React from "react";
import { MissionState } from '../../utils/HelperUtils.ts';
import 'react-toastify/dist/ReactToastify.css';

export interface ContextMenuProps {
  showMenu: boolean;
  position: { top: number; left: number };
  options: MissionState[];
  currentItem?: string;
  handleClick: (a: string, e: React.MouseEvent<HTMLDivElement, MouseEvent>) => void;
}

export const ContextMenu: React.FC<ContextMenuProps> = ({
                                                          showMenu,
                                                          position,
                                                          options,
                                                          currentItem,
                                                          handleClick,
                                                        }) => {

  if (!showMenu) return null;

  return (
    <div
      className="flex flex-col bg-gradient-to-tr border border-gray-300 rounded-md shadow-md absolute z-10 h-100 min-w-100 w-200"
      style={{ top: position.top, left: position.left }}
    >
      <div className="font-bold border-gray-50  p-5">Move to</div>
      {options.map(option => <div key={option} className="border-b-2 border-gray-50 hover:bg-amber-200 hover:cursor-pointer py-2 p-5 " onClick={(e) => handleClick(currentItem as string, e)}>{option}</div>)}
    </div>
  );
};