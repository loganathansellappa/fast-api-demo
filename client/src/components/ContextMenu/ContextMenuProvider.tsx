import React, { useCallback, useMemo, useState } from 'react';
import { ContextMenu } from "./ContextMenu";
import { MissionState, MutationType, showErrorToast, showSuccessToast } from '../../utils/HelperUtils.ts';
import { MissionControlInput, useMissionControlUpdate } from '../../hooks/useMissionControlMutation.ts';
export interface ContextMenuProviderProps {
  children: React.ReactNode;
  options: MissionState[],
}

export const ContextMenuProvider: React.FC<ContextMenuProviderProps> = ({
                                                                          children,
                                                                          options,
                                                                        }) => {
  const [contextMenu, setContextMenu] = useState<{
    show: boolean;
    top: number;
    left: number;
    item?: number;
  }>({
    show: false,
    top: 0,
    left: 0,
    item: 0,
  });
  const {  mutate, isError, isSuccess} = useMissionControlUpdate();


  const handleClick = useCallback((currentItem: number, e: React.MouseEvent<HTMLDivElement, MouseEvent>) => {
    const target = e.target as HTMLElement;
    const currentState = target.innerHTML as MissionState;
    const requestData: MissionControlInput = {
      id: currentItem as number,
      missionState: currentState,
      mutationType: MutationType.PUT,
    }
    mutate(requestData);
  },[mutate])


  if(isError) {
    showErrorToast('Mission Control state update failed', `missionControlStateUpdateFailed$`);
  }

  if(isSuccess) {
    showSuccessToast('Mission Control state updated', `missionControlStateUpdateSuccess`);
  }

  const menuOptions = useMemo(() => Object.values(options).map(state => state),[options]);
  const handleContextMenu = (e: React.MouseEvent<HTMLDivElement>) => {
    e.preventDefault();
    e.stopPropagation();
    const element = e.target as HTMLElement;
    setContextMenu({ show: true, top: e.clientY, left: e.clientX, item: element.dataset.itemId as unknown as number});
  };

  const handleContextMenuClose = () => {
    setContextMenu({ show: false, top: 0, left: 0, item: 0});
  };

  return (
    <div onContextMenu={handleContextMenu} onClick={handleContextMenuClose}>
      {children}
      <ContextMenu
        showMenu={contextMenu.show}
        position={{ top: contextMenu.top, left: contextMenu.left }}
        currentItem={contextMenu.item as number}
        options={menuOptions}
        handleClick={handleClick}
      />
    </div>
  );
};
