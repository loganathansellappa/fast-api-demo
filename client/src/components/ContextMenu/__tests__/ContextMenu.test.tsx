import { render } from '@testing-library/react';
import { ContextMenu, ContextMenuProps } from '../ContextMenu.tsx';
import { MissionState } from '../../../utils/HelperUtils.ts';

describe("ContextMenu", () => {
  let element: { container: HTMLElement };
  const clickHandler = jest.fn();
  beforeEach(() => {
    const props: ContextMenuProps = {
      showMenu: true,
      currentItem: 1,
      handleClick: clickHandler,
      position: { top: 0, left: 0 },
      options: Object.values(MissionState).map(state => state)
    };
    element = render(<ContextMenu {...props} />);
  });

  test("renders ContextMenu component", () => {
    expect(element.container.firstChild).toMatchSnapshot();
  });

  test("displays menu options", () => {
    expect(element.container.textContent).toContain(MissionState.PreFlight);
    expect(element.container.textContent).toContain(MissionState.Flight);
    expect(
      element.container.getElementsByClassName("context-menu-item")!.length,
    ).toBe(3);
  });

  test("trigger click on menu click", () => {
    const menuItem = element.container.getElementsByClassName("context-menu-item")[0] as HTMLElement;
    menuItem.click();
    expect(clickHandler).toBeCalledTimes(1);
  });
});