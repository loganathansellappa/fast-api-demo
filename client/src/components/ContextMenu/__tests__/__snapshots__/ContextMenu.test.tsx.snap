// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContextMenu renders ContextMenu component 1`] = `
<div
  class="flex flex-col bg-gradient-to-tr border border-gray-300 rounded-md shadow-md absolute z-10 h-100 min-w-100 w-200 context-menu"
  style="top: 0px; left: 0px;"
>
  <div
    class="font-bold border-gray-50  p-5"
  >
    Move to
  </div>
  <div
    class="border-b-2 border-gray-50 hover:bg-amber-200 hover:cursor-pointer py-2 p-5 context-menu-item"
  >
    Pre-flight
  </div>
  <div
    class="border-b-2 border-gray-50 hover:bg-amber-200 hover:cursor-pointer py-2 p-5 context-menu-item"
  >
    In-Flight
  </div>
  <div
    class="border-b-2 border-gray-50 hover:bg-amber-200 hover:cursor-pointer py-2 p-5 context-menu-item"
  >
    Post-Flight
  </div>
</div>
`;
